### PR TITLE
decouple criterias from pruner

### PR DESCRIPTION
**Pull request by WeiweiZhang1**
## Type of Change

architecture change、feature、bug fix and others
API not changed

## Description

Adjust the pytorch-pruner architecture, decouple criterias from pruner.
Add progressive pruner.
Add Layer Minimum Sparsity Feature.
Add yolo-coco example.

detail description

JIRA ticket: https://jira.devtools.intel.com/browse/ILITV-2344

## How has this PR been tested?
Pre-CI

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Review by WeiweiZhang1 — COMMENTED** [review]
remove development pruner in criterias.py

File: neural_compressor/experimental/pytorch_pruner/pruning.py [hunk 1]
```diff
@@ -96,68 +111,65 @@ def get_sparsity_ratio(self):
...
         for pruner in self.pruners:
             pruner.on_epoch_end()
 
+    # @_call_pruners
     def on_train_end(self):
-        """Functions called in the end of training."""
         for pruner in self.pruners:
             pruner.on_train_end()
 
+        # @_call_pruners
```
Comment (WeiweiZhang1) [inline neural_compressor/experimental/pytorch_pruner/pruning.py]: Adjust indent

File: neural_compressor/experimental/pytorch_pruner/criterias.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    def on_after_optimizer_step(self):
+        with torch.no_grad():
+            for key in self.modules.keys():
+                p = self.modules[key].weight
+                self.scores[key] *= self.alpha
+                self.scores[key] += self.beta * torch.abs(p * p.grad)
+
+
+"""
+the following pruners are under development
```
[the remark this answers is not in the document]
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/criterias.py]: done

File: neural_compressor/experimental/pytorch_pruner/patterns.py [hunk 3]
```diff
@@ -20,23 +19,21 @@
...
     """Class decorator used to register a Pattern subclass to the registry.
-
-    Decorator function used before a Pattern subclasses. 
+    Decorator function used before a Pattern subclasses.
     Make sure that this Pattern class can be registered in PATTERNS.
-
     Args:
         cls (class): The class of register.
         name: A string. Define the pattern type which will be included in a pruning process.
-
```
[the remark this answers is not in the document]
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/patterns.py:36]: done

File: neural_compressor/experimental/pytorch_pruner/pruning.py (hunk 1 above)
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/pruning.py]: done

File: neural_compressor/experimental/pytorch_pruner/patterns.py [hunk 4]
```diff
@@ -493,13 +578,15 @@ def reshape_orig_to_pattern(self, data: torch.Tensor, key):
...
-        reshape the data [s1/N,s2/M] to [s1,s2], also permute dims for conv layer
+        Reshape the data [s1/N,s2/M] to [s1,s2], also permute dims for conv layer.
+
         Args:
             data:
             key:
             orig_shape:
             
         Returns:
+            data:
```
Comment (WeiweiZhang1) [inline neural_compressor/experimental/pytorch_pruner/patterns.py:589]: Adjust docstring of returns

File: neural_compressor/experimental/pytorch_pruner/patterns.py [hunk 5]
```diff
@@ -69,17 +76,29 @@ def get_pattern(config, modules):
...
         is_global: A bool. Whether the pruning take global pruning option.
                    Global pruning means that all pruning layers are gathered to calculate pruning criteria.
                    Local pruning, on the contrast, means that pruning layers are to calculate criteria individually.
+        keep_mask_layers:A dict. the layers whose mask will not be updated
+        invalid_layers: the layers whose shape don't fit the patten
+        modules: torch neural network modules, which will be pruned with the pattern
+        config: A config dict object. Contains all the information including the pattern's.
+        max_sparsity_ratio_per_layer: A float. The maximum sparsity that one layer could reach
+        min_sparsity_ratio_per_layer: A float. The minimum sparsity that one layer could reach
+        target_sparsity: A float. The sparsity ratio of the modules will be reached after pruning,
```
Comment (WeiweiZhang1) [inline neural_compressor/experimental/pytorch_pruner/patterns.py:96]: ... after pruning,
->
... after pruning.

File: neural_compressor/experimental/pytorch_pruner/pruner.py [hunk 6]
```diff
@@ -115,87 +132,248 @@ def __init__(self, config, modules):
...
+        current_target_sparsity_ratio = self.scheduler.update_sparsity_ratio(self.target_sparsity_ratio,
+                                                                             self.completed_pruned_cnt,
+                                                                             self.total_prune_cnt, self.masks,
+                                                                             self.init_sparsity_ratio)
+        logger.info(f"current target ratio is {current_target_sparsity_ratio}")
+
+        self.completed_pruned_cnt += 1
+        if self.criteria.scores == {}:
+            return
+        self.masks = self.pattern.get_masks(self.criteria.scores, current_target_sparsity_ratio, self.masks,
```
Comment (WeiweiZhang1) [inline neural_compressor/experimental/pytorch_pruner/pruner.py:297]: Fix the problem of passing references to get_masks func

Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/pruner.py:297]: done

File: neural_compressor/experimental/pytorch_pruner/patterns.py (hunk 5 above)
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/patterns.py:96]: done

File: neural_compressor/experimental/pytorch_pruner/patterns.py (hunk 4 above)
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/patterns.py:589]: done

File: neural_compressor/experimental/pytorch_pruner/criterias.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        """Calculate and store the pruning scores of pruning modules at the beginning of a step."""
+        pass
+
+    def on_after_optimizer_step(self):
+        """Calculate and store the pruning scores of pruning modules after the optimizer step."""
+        pass
+
+
+@register_criteria('magnitude')
+class MagnitudeCriteria(Criteria):
```
Comment (ftian1) [inline neural_compressor/experimental/pytorch_pruner/criterias.py:68]: spelling check. Criteria is plural. criterion is singular

File: neural_compressor/experimental/pytorch_pruner/criterias.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    def on_step_begin(self):
+        """Calculate and store the pruning scores of pruning modules at the beginning of a step."""
+        pass
+
+    def on_after_optimizer_step(self):
+        """Calculate and store the pruning scores of pruning modules after the optimizer step."""
+        pass
+
+
+@register_criteria('magnitude')
```
Comment (ftian1) [inline neural_compressor/experimental/pytorch_pruner/criterias.py:67]: should be register_criterion(...)

File: neural_compressor/experimental/pytorch_pruner/criterias.py (hunk 7 above)
Comment (ftian1) [reply neural_compressor/experimental/pytorch_pruner/criterias.py:68]: and pls modify base class name

File: neural_compressor/experimental/pytorch_pruner/criterias.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+    def on_after_optimizer_step(self):
+        """Calculate and store the pruning scores based on gradient criterion."""
+        with torch.no_grad():
+            for key in self.modules.keys():
+                p = self.modules[key].weight
+                self.scores[key] = torch.abs(p.grad)
+
+@register_criteria('snip')
+class SnipCriteria(Criteria):
```
Comment (ftian1) [inline neural_compressor/experimental/pytorch_pruner/criterias.py:120]: confirm with Wenhua which words are more proper for use

File: neural_compressor/experimental/pytorch_pruner/prune_utils.py [hunk 10]
```diff
@@ -189,7 +221,7 @@ def process_config(config):
     return process_and_check_config(val)
 
 
-def parse_to_prune(model, config):
+def parse_to_prune(config, model):
```
Comment (ftian1) [inline neural_compressor/experimental/pytorch_pruner/prune_utils.py:224]: why the parameter's order is changed? any special reason?

File: neural_compressor/experimental/pytorch_pruner/pruner.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-"""pruner module."""
+"""Pruner."""
```
Comment (ftian1) [inline neural_compressor/experimental/pytorch_pruner/pruner.py:1]: weiwei to confirm w/ wenhua about the directory naming

File: examples/pytorch/object_detection/yolo_v5/pruning/pytorch_pruner/eager/README.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+```yaml
+version: 1.0
+
+model:
+  name: "yolov5s6"
+  framework: "pytorch"
+
+pruning:
+  approach:
+    weight_compression_pytorch:
```
Comment (ftian1) [inline examples/pytorch/object_detection/yolo_v5/pruning/pytorch_pruner/eager/README.md:20]: should be "weight_compression" but not "weight_compression_pytorch".

File: examples/pytorch/object_detection/yolo_v5/pruning/pytorch_pruner/eager/README.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
+# Pytorch Pruner
+## Intro
+[**Pytorch Pruner**](https://github.com/intel/neural-compressor/tree/master/neural_compressor/experimental/pytorch_pruner) is an INC build-in API which supports a wide range of pruning algorithms, patterns as well as pruning schedulers. Features below are currently supported:
+> algorithms: magnitude, snip, snip-momentum\
+> patterns: NxM, N:M\
+> pruning schedulers: iterative pruning scheduler, one-shot pruning scheduler.
+
+## Usage
+### Write a config yaml file
+Pytorch pruner is developed based on [pruning](https://github.com/intel/neural-compressor/blob/master/neural_compressor/experimental/pruning.py), therefore most usages are identical. Our API reads in a yaml configuration file to define a Pruning object. Here is an bert-mini example of it:
```
Comment (ftian1) [inline examples/pytorch/object_detection/yolo_v5/pruning/pytorch_pruner/eager/README.md:10]: PyTorch but not Pytorch

File: neural_compressor/experimental/pytorch_pruner/pruner.py (hunk 11 above)
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/pruner.py:1]: INC has a pruner, so use pytorch_pruner for now

File: neural_compressor/experimental/pytorch_pruner/prune_utils.py (hunk 10 above)
Comment (WeiweiZhang1) [reply neural_compressor/experimental/pytorch_pruner/prune_utils.py:224]: In order to unify the order of the two formal parameters in the internal functions, the change does not affect the pruner API.

**Comment by ftian1** [comment]
@WeiweiZhang1 shall you close this PR?
